File: 28-08-2024_Sorting_Elements_of_an_Array_by_Frequency.py
from collections import Counter

class Solution:
    def sortByFreq(self, arr):
        freq_map = Counter(arr)
        sorted_items = sorted(freq_map.items(), key=lambda x: (-x[1], x[0]))
        
        ans = []
        for num, freq in sorted_items:
            ans.extend([num] * freq)
        
        return ans

File: 28-08-2024_Sorting_Elements_of_an_Array_by_Frequency.java
class Solution {

    public static ArrayList<Integer> sortByFreq(int[] arr) {
        Map<Integer, Integer> freqMap = new HashMap<>();
        for (int num : arr) {
            freqMap.put(num, freqMap.getOrDefault(num, 0) + 1);
        }

        ArrayList<Map.Entry<Integer, Integer>> entryList = new ArrayList<>(freqMap.entrySet());

        entryList.sort((a, b) -> {
            if (a.getValue().equals(b.getValue())) {
                return a.getKey() - b.getKey();
            }
            return b.getValue() - a.getValue();
        });

        ArrayList<Integer> ans = new ArrayList<>();
        for (Map.Entry<Integer, Integer> entry : entryList) {
            for (int i = 0; i < entry.getValue(); i++) {
                ans.add(entry.getKey());
            }
        }
        return ans;
    }
}
